kernel/src/net/socket/unix/ns/abs.rs

```diff
@@ -0,0 +1,122 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::btree_map::Entry, format};
+
+use keyable_arc::KeyableArc;
+
+use crate::prelude::*;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct AbstractHandle(KeyableArc<[u8]>);
+
+impl AbstractHandle {
+    fn new(name: Arc<[u8]>) -> Self {
+        Self(KeyableArc::from(name))
+    }
+
+    pub fn name(&self) -> Arc<[u8]> {
+        self.0.clone().into()
+    }
+}
+
+impl Drop for AbstractHandle {
+    fn drop(&mut self) {
+        HANDLE_TABLE.remove(self.name());
+    }
+}
+
+static HANDLE_TABLE: HandleTable = HandleTable::new();
+
+struct HandleTable {
+    handles: RwLock<BTreeMap<Arc<[u8]>, Weak<AbstractHandle>>>,
+}
```

Comment:
I don't think these two will work:
 - `BTreeSet<KeyableWeak<AbstractHandle>` simply won't compile, since we can't implement `Ord` for `Weak<_>` due to fallible `Weak::upgrade`.
 - `BTreeSet<KeyableArc<AbstractHandle>` will cause a memory leak, since `AbstractHandle::drop` can never be called if the reference count does not reach zero.

Reply:
My concern is that manually managing the reference count is hard to check its correctness. Is there any possible solution to avoid this problem?

> BTreeSet<KeyableWeak<AbstractHandle> simply won't compile, since we can't implement Ord for Weak<_> due to fallible Weak::upgrade

Our `KeyableWeak` can be used as a key because it uses address as key and does implement `Ord`.

```rust
impl<T: ?Sized> Ord for KeyableWeak<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        self.0
            .as_ptr()
            .cast::<()>()
            .cmp(&other.0.as_ptr().cast::<()>())
    }
}
```

Reply:
> manually managing the reference count

I think no one is trying to mangle the reference count manually? The only thing being done manually here is to remove the entry from the map after the reference count reaches zero, and I don't think there's a way to do that automatically.

> Our KeyableWeak can be used as a key because it uses address as key and does implement Ord.

That's true, but we need to find it by name, not by address.

Reply:
Thanks for your explanation! Then the current form is acceptable for me.